File: logger-examples/python-fastapi/logger.py
import requests
import os
import json
from datetime import datetime
from dotenv import load_dotenv
from fastapi import Request
from user_agents import parse as ua_parse

load_dotenv()


def info(message, request):
    log(request, 'INFO', message)


def warn(message, request):
    log(request, 'WARN', message)


def error(message, request):
    log(request, 'ERROR', message)


def log(req: Request, log_level, log_message):
    user_agent = ua_parse(req.headers['user-agent'])
    data = {
        "event_ts": datetime.now().isoformat(),
        "ip_address": req.client.host,
        "city": 'unknown',
        "country": 'unknown',
        "region": 'unknown',
        "latitude": 'unknown',
        "longitude": 'unknown',
        "protocol": req.headers['x-forwarded-proto'] if req.headers.get('x-forwarded-proto') else 'unknown',
        "method": req.method,
        "host": req.headers['host'],
        "url": f'{req.url.scheme}//:{req.headers["host"]}{req.url.path}',
        "headers": ','.join(req.headers.keys()),
        "useragent": req.headers['user-agent'] if req.headers.get('user-agent') else 'unknown',
        "referer": req.headers['referer'] if req.headers.get('referer') else 'unknown',
        "acceptencoding": req.headers['accept-encoding'] if req.headers.get('accept-encoding') else 'unknown',
        "acceptlanguage": req.headers['accept-language'] if req.headers.get('accept-language') else 'unknown',
        "acceptcharset": req.headers['accept-charset'] if req.headers.get('accept-charset') else 'unknown',
        "origin": req.headers['origin'] if req.headers.get('origin') else 'unknown',
        "xforwaredforip": req.headers['x-forwarded-for'] if req.headers.get('x-forwarded-for') else 'unknown',
        "connection": req.headers['connection'] if req.headers.get('connection') else 'unknown',
        "cachecontrol": req.headers['cache-control'] if req.headers.get('cache-control') else 'unknown',
        "contenttype": req.headers['content-type'] if req.headers.get('content-type') else 'unknown',
        "from": req.headers['from'] if req.headers.get('from') else 'unknown',
        "via": req.headers['via'] if req.headers.get('via') else 'unknown',
        "contentlength": req.headers['content-length'] if req.headers.get('content-length') else 'unknown',
        "isbot": 0,
        "browsername": user_agent.browser.family if user_agent.browser.family else 'unknown',
        "browserversion": user_agent.browser.version_string if user_agent.browser.version_string else 'unknown',
        "devicemodel": user_agent.device.model if user_agent.device.model else 'unknown',
        "devicetype": user_agent.device.family if user_agent.device.family else 'unknown',
        "devicevendor": user_agent.device.brand if user_agent.device.brand else 'unknown',
        "enginename": 'unknown',
        "engineversion": 'unknown',
        "osname": user_agent.os.family if user_agent.os.family else 'unknown',
        "osversion": user_agent.os.version_string if user_agent.os.version_string else 'unknown',
        "cpuarchitecture": 'unknown',
        "log_level": log_level,
        "log_message": log_message,
    }
    send_to_tinybird(data)


def send_to_tinybird(data):
    r = requests.post('https://api.tinybird.co/v0/events?name=logs',
                      params={
                          'token': os.getenv('TINYBIRD_TOKEN'),
                      },
                      data=json.dumps(data))
    return True

File: logger-examples/typescript-nextjs/logger.ts
import { NextRequest, userAgent } from 'next/server'

export class logger {
  static async info(message: string, request: NextRequest) {
    this.log(request, 'INFO', message)
  }

  static async warn(message: string, request: NextRequest) {
    this.log(request, 'WARN', message)
  }

  static async error(message: string, request: NextRequest) {
    this.log(request, 'ERROR', message)
  }

  static async log(req: NextRequest, logLevel: string, logMessage: string) {
    const { isBot, browser, device, engine, os, cpu } = userAgent(req)

    const data = {
      event_ts: new Date().toISOString(),
      ip_address: req.ip,
      city: req.geo?.city ?? 'unknown',
      country: req.geo?.country ?? 'unknown',
      region: req.geo?.region ?? 'unknown',
      latitude: req.geo?.latitude ?? 'unknown',
      longitude: req.geo?.longitude ?? 'unknown',
      protocol: req.headers.get('x-forwarded-proto') ?? 'unknown',
      method: req.method,
      host: req.headers.get('host'),
      url: req.url,
      headers: Array.from(req.headers.keys()).join(','),
      useragent: req.headers.get('user-agent') ?? 'unknown',
      referer: req.headers.get('referer') ?? 'unknown',
      acceptencoding: req.headers.get('accept-encoding') ?? 'unknown',
      acceptlanguage: req.headers.get('accept-language') ?? 'unknown',
      acceptcharset: req.headers.get('accept-charset'),
      origin: req.headers.get('origin') ??'unknown',
      xforwaredforip: req.headers.get('x-forwarded-for') ?? 'unknown',
      connection: req.headers.get('connection') ?? 'unknown',
      cachecontrol: req.headers.get('cache-control') ?? 'unknown',
      contenttype: req.headers.get('content-type') ?? 'unknown',
      from: req.headers.get('from') ?? 'unknown',
      via: req.headers.get('via') ?? 'unknown',
      contentlength: req.headers.get('content-length'),
      isbot: isBot,
      browsername: browser.name ?? 'unknown',
      browserversion: browser.version ?? 'unknown',
      devicemodel: device.model ?? 'unknown',
      devicetype: device.type ?? 'unknown',
      devicevendor: device.vendor ?? 'unknown',
      enginename: engine.name ?? 'unknown',
      engineversion: engine.version ?? 'unknown',
      osname: os.name ?? 'unknown',
      osversion: os.version ?? 'unknown',
      cpuarchitecture: cpu.architecture ?? 'unknown',
      log_level: logLevel,
      log_message: logMessage,
    }

    await this.sendToTinybird(data)
  }

  static async sendToTinybird(data: Record<string, unknown>) {
    await fetch(
      `https://api.tinybird.co/v0/events?name=logs`,
      {
        method: 'POST',
        body: JSON.stringify(data),
        headers: { Authorization: `Bearer ${process.env.TINYBIRD_TOKEN}` },
      }
    )
  }
}
